packages/devtools_app/lib/src/ui/hover.dart

```diff
@@ -358,74 +368,139 @@ class _HoverCardTooltipState extends State<HoverCardTooltip> {
+      final hoverCardDataCompleter =
+          _hoverCardDataCompleter(hoverCardDataFuture);
+      await Future.any([
+        _timeoutCompleter(asyncTimeout).future,
+        hoverCardDataCompleter.future,
+      ]);
+
+      if (hoverCardDataCompleter.isCompleted) {
+        final data = await hoverCardDataCompleter.future;
+        // If we get no data back, then don't show a hover card.
+        if (data == null) return;
+        // Otherwise, show a hover card immediately.
+        return _setHoverCardFromData(
+          data,
+          context: context,
           event: event,
-          isHoverStale: () =>
-              !_hoverCardController.isHoverCardStillActive(spinnerHoverCard),
         );
+      }
+    }
+    // The data on the card is fetched asynchronously, so show a spinner
+    // while we wait for it.
+    spinnerHoverCard = HoverCard.fromHoverEvent(
+      context: context,
+      contents: const CenteredCircularProgressIndicator(),
+      width: HoverCardTooltip.defaultHoverWidth,
+      event: event,
+      hoverCardController: _hoverCardController,
+    );
 
-        if (!_hoverCardController.isHoverCardStillActive(spinnerHoverCard)) {
-          // The hovercard became stale while fetching it's data. So it should
-          // no longer be shown.
-          return;
-        }
-        if (hoverCardData == null) {
-          // No data was provided so remove the spinner
-          _removeHoverCard(spinnerHoverCard);
-          return;
-        }
-      } else {
-        assert(widget.generateHoverCardData != null);
+    _setHoverCard(
+      spinnerHoverCard,
+    );
 
-        hoverCardData = widget.generateHoverCardData!(event);
-      }
+    // The spinner is showing, we can now generate the HoverCardData
+    final hoverCardData = await hoverCardDataFuture;
 
-      if (!mounted) return;
-
-      if (hoverCardData.position == HoverCardPosition.cursor) {
-        _setHoverCard(
-          HoverCard.fromHoverEvent(
-            context: context,
-            title: hoverCardData.title,
-            contents: hoverCardData.contents,
-            width: hoverCardData.width,
-            event: event,
-            hoverCardController: _hoverCardController,
-          ),
-        );
-      } else {
-        _setHoverCard(
-          HoverCard(
-            context: context,
-            title: hoverCardData.title,
-            contents: hoverCardData.contents,
-            width: hoverCardData.width,
-            position: _calculateTooltipPosition(hoverCardData.width),
-            hoverCardController: _hoverCardController,
-          ),
-        );
-      }
+    if (!_hoverCardController.isHoverCardStillActive(spinnerHoverCard)) {
+      // The hovercard became stale while fetching it's data. So it should
+      // no longer be shown.
+      return;
+    }
+    if (hoverCardData == null) {
+      // No data was provided so remove the spinner
+      _removeHoverCard(spinnerHoverCard);
+      return;
+    }
```

Comment:
I think we might need to post the hoverCard data in a hovercard here
hoverCardData
otherwise if we wait long enough to get the spinner, the official hovercard won't ever show

Reply:
Oops! Thanks for catching that